### PR TITLE
Clean up PVCs that are not owned by VRG but present in protected pvcs list

### PR DESCRIPTION
**Pull request by asn1809**
This helps in solving https://github.com/RamenDR/ramen/issues/1335

### CODE REVIEW

**Comment by ShyamsundarR** [comment]
Needs envtests to the added. [This](https://github.com/RamenDR/ramen/blob/f36cc46ba801cc1930264e240c2327092ac9cf56/controllers/vrg_volrep_test.go#L355) section hase some tests around this functionality that either is not validating for the current behavior or does not test the same and may need modifications and addition appropriately.

**Comment by ShyamsundarR** [comment]
[`pvcStatusDeleteIfPresent`](https://github.com/RamenDR/ramen/blob/f36cc46ba801cc1930264e240c2327092ac9cf56/controllers/vrg_status_pvcs.go#L22) seems to be doing the same when invoked for a PVC that should no longer be protected. Can you check if that function is not doing the required operation correctly?

Also, instead of clearing the PVCs from the protectedPVCs list at the end of `pvcsDeselectedUnprotect` it would be easier to do it how (or what seems to be intended) per PVC that is unprotected in `pvcStatusDeleteIfPresent`

**Comment by pdumbre** [comment]
In function 'pvcsDeselectedUnprotect' , PVC list is retrieved by 'listPVCsOwnedByVrg' method. Can we consider PVCs not having owner labels for 'pvcsOwned' list here?  In the subsequent for loop, PVCs currently filtered out by recipe selector and still present in 'protectedPVC' list of vrg are cleaned up both from S3 store and vrg status as well.

File: controllers/volumereplicationgroup_controller.go [hunk 1]
```diff
@@ -966,9 +966,59 @@ func (v *VRGInstance) pvcsDeselectedUnprotect() error {
...
+
+	for _, volSyncPVC := range v.volSyncPVCs {
+		if pvc.Name == volSyncPVC.Name && pvc.Namespace == volSyncPVC.Namespace {
+			protectedPVCsFiltered = append(protectedPVCsFiltered, pvc)
+		}
+	}
+
+	return protectedPVCsFiltered
+}
+
```
Comment (ShyamsundarR) [inline controllers/volumereplicationgroup_controller.go:1021]: We can potentially simplify the loops reusing the object maps that are created. Here is a suggestion, PTAL and see if this works for you.

```suggestion
func (v *VRGInstance) cleanUpProtectedPVCs(log logr.Logger, pvcsVr, pvcsVs map[client.ObjectKey]corev1.PersistentVolumeClaim) {
	if !v.ramenConfig.VolumeUnprotectionEnabled {
		log.Info("Volume unprotection disabled")

		return
	}

	if v.instance.Spec.Async != nil && !VolumeUnprotectionEnabledForAsyncVolRep {
		log.Info("Volume unprotection disabled for async mode")

		return
	}

	// clean up the PVCs that are part of protected pvcs but not in v.volReps and v.volSyncs
	protectedPVCsFiltered := make([]ramendrv1alpha1.ProtectedPVC, 0)
	for _, protectedPVC := range v.instance.Status.ProtectedPVCs {
		pvcNamespacedName := client.ObjectKey{Namespace: protectedPVC.Namespace, Name: protectedPVC.Name}

		if _, ok := pvcsVr[pvcNamespacedName]; ok {
			protectedPVCsFiltered = append(protectedPVCsFiltered, protectedPVC)

			continue
		}

		if _, ok := pvcsVs[pvcNamespacedName]; ok {
			protectedPVCsFiltered = append(protectedPVCsFiltered, protectedPVC)

			continue
		}
	}

	v.instance.Status.ProtectedPVCs = protectedPVCsFiltered
}
```

**Comment by ShyamsundarR** [comment]
> [`pvcStatusDeleteIfPresent`](https://github.com/RamenDR/ramen/blob/f36cc46ba801cc1930264e240c2327092ac9cf56/controllers/vrg_status_pvcs.go#L22) seems to be doing the same when invoked for a PVC that should no longer be protected. Can you check if that function is not doing the required operation correctly?
> 
> Also, instead of clearing the PVCs from the protectedPVCs list at the end of `pvcsDeselectedUnprotect` it would be easier to do it how (or what seems to be intended) per PVC that is unprotected in `pvcStatusDeleteIfPresent`

Responding to this after discussing with @Annaraya-Narasagond for PVCs that were not Bound they do not have the owner label, but are in the protectedPVCs list and hence need garbage collection.

File: controllers/volumereplicationgroup_controller.go (hunk 1 above)
Comment (asn1809) [reply controllers/volumereplicationgroup_controller.go:1009]: Done

**Comment by ShyamsundarR** [comment]
> @ShyamsundarR @Annaraya-Narasagond : In function 'pvcsDeselectedUnprotect' , PVC list is retrieved by 'listPVCsOwnedByVrg' method. Can we consider PVCs not having owner labels for 'pvcsOwned' list here? In the subsequent for loop, PVCs currently filtered out by recipe selector and still present in 'protectedPVC' list of vrg are cleaned up both from S3 store and vrg status as well.

This maybe an overkill as we would be listing and processing all PVCs in said namespace. It does resolve the issue, but increases the scope of processing that we can avoid.

**Comment by ShyamsundarR** [comment]
> Needs envtests to the added. [This](https://github.com/RamenDR/ramen/blob/f36cc46ba801cc1930264e240c2327092ac9cf56/controllers/vrg_volrep_test.go#L355) section hase some tests around this functionality that either is not validating for the current behavior or does not test the same and may need modifications and addition appropriately.

@asn1809 Assume envtests are still pending as part of this PR. Current changes LGTM!

File: controllers/volumereplicationgroup_controller.go [hunk 2]
```diff
@@ -966,9 +966,47 @@ func (v *VRGInstance) pvcsDeselectedUnprotect() error {
...
+		if _, ok := pvcsVr[pvcNamespacedName]; ok {
+			protectedPVCsFiltered = append(protectedPVCsFiltered, protectedPVC)
+
+			continue
+		}
+
+		if _, ok := pvcsVs[pvcNamespacedName]; ok {
+			protectedPVCsFiltered = append(protectedPVCsFiltered, protectedPVC)
+
+			continue
```
Comment (ELENAGER) [inline controllers/volumereplicationgroup_controller.go:1003]: I think, continue is not needed here

**Review by ShyamsundarR — APPROVED** [review]

**Pull request merged**